staging/src/k8s.io/kube-aggregator/pkg/apis/apiregistration/types.go

```diff
@@ -53,13 +53,23 @@ type APIServiceSpec struct {
 	// CABundle is a PEM encoded CA bundle which will be used to validate an API server's serving certificate.
 	CABundle []byte
 
-	// Priority controls the ordering of this API group in the overall discovery document that gets served.
-	// Client tools like `kubectl` use this ordering to derive preference, so this ordering mechanism is important.
-	// Values must be between 1 and 1000
-	// The primary sort is based on priority, ordered lowest number to highest (10 before 20).
+	// GroupPriority controls the ordering of the group with respect to other groups. The lowest of all GroupPriorities
+	// for the same group is used.  Must be greater than zero.
+	// The primary sort is based on GroupPriority, ordered lowest number to highest (10 before 20).
+	// The secondary sort is based on the VersionPriority, ordered lowest number to highest.
+	// The tertiary sort is based on the alphabetical comparison of the name of the object.  (v1.bar before v1.foo)
+	// We'd recommend something like: *.k8s.io (except extensions) at 1000, extensions at 2000
+	// PaaSes (OpenShift, Deis) are recommended to be in the 3000s
+	GroupPriority int32
```

Comment:
> But what does it mean if a user puts two versions in the same group with different group priorities?

The group is as important as its most important group version 

> What if you want to change the group priority? You can't change every groupversion's groupPriority atomically.

Set the group priority on the APIService objects for that group. If you are making the group more important, as soon as the first change is observed by the aggregator, the group increases in importance. The order in discovery already depends on observation by the aggregator watching the changes to the registration objects... there is nothing atomic about it. 

> I'm really against API designs with state spaces that include a lot of nonsensical configurations.

The current priority field lets you express interleaved priority between multiple groups and versions that does not actually control the final priority of a group version

Separating the fields does not make that problem worse, and isolates the version priority so that it can be specified without worrying about side effects